drivers/counter/counter_ambiq_timer.c

```diff
@@ -163,33 +249,70 @@ static const struct counter_driver_api counter_api = {
+		counter_ambiq_get_value(dev_##n, &now);                                            \
+		if (data_##n->callback) {                                                          \
+			data_##n->callback(dev_##n, 0, now, data_##n->user_data);                  \
+		}                                                                                  \
+	}
+
 static void counter_ambiq_isr(void *arg)
 {
+	uint32_t now = 0;
+
+#if defined(CONFIG_SOC_SERIES_APOLLO3X)
+	ARG_UNUSED(arg);
+
+	DT_INST_FOREACH_STATUS_OKAY(APOLLO3_HANDLE_SHARED_TIMER_IRQ)
+#else
 	const struct device *dev = (const struct device *)arg;
 	struct counter_ambiq_data *data = dev->data;
-	uint32_t now = 0;
+	const struct counter_ambiq_config *cfg = dev->config;
 
-	am_hal_timer_interrupt_clear(AM_HAL_TIMER_MASK(0, AM_HAL_TIMER_COMPARE1));
+	am_hal_timer_interrupt_clear(AM_HAL_TIMER_MASK(cfg->instance, AM_HAL_TIMER_COMPARE1));
 	counter_ambiq_get_value(dev, &now);
 
 	if (data->callback) {
 		data->callback(dev, 0, now, data->user_data);
 	}
+#endif
 }
 
+#if defined(CONFIG_SOC_SERIES_APOLLO3X)
+/* Apollo3 counters share the same irq number, connect irq here will cause build error, so we
+ * leave this function blank here and do it in counter_irq_config_func
+ */
+#define AMBIQ_COUNTER_CONFIG_FUNC(idx) static void counter_irq_config_func_##idx(void){};
+#else
+#define AMBIQ_COUNTER_CONFIG_FUNC(idx)                                                             \
+	static void counter_irq_config_func_##idx(void)                                            \
+	{                                                                                          \
+		NVIC_ClearPendingIRQ(DT_INST_IRQN(idx));                                           \
+		IRQ_CONNECT(DT_INST_IRQN(idx), DT_INST_IRQ(idx, priority), counter_ambiq_isr,      \
+			    DEVICE_DT_INST_GET(idx), 0);                                           \
+		irq_enable(DT_INST_IRQN(idx));                                                     \
+	};
+#endif
+
 #define AMBIQ_COUNTER_INIT(idx)                                                                    \
-                                                                                                   \
+	static void counter_irq_config_func_##idx(void);                                           \
 	static struct counter_ambiq_data counter_data_##idx;                                       \
-                                                                                                   \
 	static const struct counter_ambiq_config counter_config_##idx = {                          \
+		.instance = (DT_INST_REG_ADDR(idx) - COUNTER_BASE) / DT_INST_REG_SIZE(idx),        \
 		.counter_info = {.max_top_value = UINT32_MAX,                                      \
-				 .freq = 6000000,                                                  \
+				 .freq = COUNTER_FREQ,                                             \
 				 .flags = COUNTER_CONFIG_INFO_COUNT_UP,                            \
```

Comment:
Can we have a clock-frequency property for "ambiq,counter.yaml" and define the value in the dts of SOC? And the COUNTER_BASE should be equal to DT_INST_REG_ADDR(0), right? I don't think it is necessary to define the macro in the driver.